### PR TITLE
[gsi_interface] Introduces canAccessScopes and userDataEvents.

### PR DESCRIPTION
**Pull request by ditman**
Introduces: `canAccessScopes` method and `userDataEvents` stream to the platform_interface of the google_sign_in plugin.

Default implementations:

* canAccessScopes => throw not implemented.
* userDataEvents => null (currently user data events cannot be introduced asynchronously from the native side)

The changes above enable the separation of Authentication and Authorization, and asynchronous sign-in operations as needed (the web needs this now).

## Issues

* Starts landing: https://github.com/flutter/packages/pull/3461
* Enables fixing: https://github.com/flutter/flutter/issues/121781
  * Enables fixing: https://github.com/flutter/flutter/issues/121275

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart [hunk 1]
```diff
@@ -139,4 +139,22 @@ abstract class GoogleSignInPlatform extends PlatformInterface {
...
+
+  /// Determines if the current user can access all [scopes].
+  ///
+  /// Optionally, an [accessToken] can be passed for applications where a
+  /// long-lived token may be cached (like the web).
+  Future<bool> canAccessScopes(
+    List<String> scopes, {
+    String? accessToken,
+  }) async {
+    return isSignedIn();
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart:151]: Yes, I can make this throwy. I'll have to tweak the example app later so this only gets called `if (kIsWeb)` or `try/catch` its usage.

File: packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart [hunk 2]
```diff
@@ -139,4 +139,22 @@ abstract class GoogleSignInPlatform extends PlatformInterface {
...
+
+  /// Determines if the current user can access all [scopes].
+  ///
+  /// Optionally, an [accessToken] can be passed for applications where a
+  /// long-lived token may be cached (like the web).
+  Future<bool> canAccessScopes(
+    List<String> scopes, {
+    String? accessToken,
+  }) async {
+    throw UnimplementedError('canAccessScopes() has not been implmented.');
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart:151]: Typo: implemented

Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart:151]: Oops, it's in a bunch of the methods. Could you mass-fix this file?

**Review by stuartmorgan — APPROVED** [review]
LGTM with nit

File: packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart (hunk 2 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in_platform_interface/lib/google_sign_in_platform_interface.dart:151]: > Oops, it's in a bunch of the methods. Could you mass-fix this file?

Absolutely, I copy-pasted and didn't notice the typo!

**Comment by ditman** [comment]
Triggering a branch update to see if the "Windows" checks get unstuck.

**Comment by ditman** [comment]
My dearest autosubmit bot, please take a look at this PR :')

**Pull request merged**